### PR TITLE
FIX: Change data store

### PR DESCRIPTION
**Pull request by lis2**
Done
- [x] move data from PluginStore to EncryptedTopicsUser
- [x] move keys from UserCustomField to UserEncryptionKey
- [x] move data from TopicCustomField

How I tested

Create a new database with 2 users, both with encryption keys and secret message (plugin with master branch). In this scenario, migration of data was tested. 
- [x] change plugin branch to `change-data-store` and run migration
- [x] ensure that after migration new tables are populated (UserEncryptionKey, EncryptedTopicsUser, EncryptedTopicsTitle)
- [x] ensure that both users can decode the private message with old paper keys
- [x] reverted 3 migrations (20200223214818,  20200130050409, 20200129230327), checkout plugin to master branch and ensure that user can decrypt message

In this scenario, check if everything works fine when adding a plugin for the first time
Create a new database with 3 users, both with encryption keys and secret message (plugin with master branch)
- [x] generate new keys and ensure that UserEncryptionKey is populated
- [x] add a few devices and remove a few devices
- [x] send a secret message to the second user
- [x] invite the third user to the message
- [x] answer as the third user
- [x] export/import paper keys

### CODE REVIEW

**Comment by lis2** [comment]
@udan11 could you review that change? I tried to test it carefully, but that changes are very risky. I will play with that tomorrow as well to ensure it is not breaking anything badly

File: plugin.rb [hunk 1]
```diff
@@ -15,28 +15,25 @@
...
-    PLUGIN_NAME          = 'discourse-encrypt'
-
-    PUBLIC_CUSTOM_FIELD  = 'encrypt_public'
-    PRIVATE_CUSTOM_FIELD = 'encrypt_private'
-    TITLE_CUSTOM_FIELD   = 'encrypted_title'
-
-    Store = PluginStore.new(PLUGIN_NAME)
+    PLUGIN_NAME = 'discourse-encrypt'
 
     def self.set_key(topic_id, user_id, key)
```
Comment (udan11) [inline plugin.rb]: Nitpick: I am wondering if we can get rid of all these methods: `set_key`, `get_key` and `del_key`. I originally introduced them because I did now want to interact with the plugin store directly (i.e. always pass plugin name or an instance of `PluginStore`, build key, etc).

File: plugin.rb [hunk 2]
```diff
@@ -237,6 +229,11 @@ class DiscourseEncrypt::Engine < Rails::Engine
       keys.each { |u, k| DiscourseEncrypt::set_key(topic_id, users[u].id, k) }
     end
 
+    if result.success? && encrypted_title = manager.args[:encrypted_title]
+      encrypt_topic_title = EncryptedTopicsTitle.find_or_initialize_by(topic_id: result.post.topic_id)
```
Comment (udan11) [inline plugin.rb]: Nitpick: Same as before, not a fan of `find_or_initialize_by`.

File: assets/javascripts/lib/discourse.js.es6 [hunk 3]
```diff
@@ -353,6 +353,9 @@ function upgradeIdentity(currentUser, passphrase, oldIdentity) {
               passphrase: exported.private
             });
 
+            currentUser.set("encrypt_public", exported.public);
```
Comment (udan11) [inline assets/javascripts/lib/discourse.js.es6]: Why is this needed?

File: app/controllers/encrypt_controller.rb [hunk 4]
```diff
@@ -55,10 +55,10 @@ def update_keys
   def delete_key
     private_id_label = params.require(:label)
 
-    data = JSON.parse(current_user.custom_fields[DiscourseEncrypt::PRIVATE_CUSTOM_FIELD]) rescue {}
+    data = JSON.parse(current_user.user_encryption_key.encrypt_private) rescue {}
     if data.delete(private_id_label)
-      current_user.custom_fields[DiscourseEncrypt::PRIVATE_CUSTOM_FIELD] = JSON.dump(data)
-      current_user.save_custom_fields
+      user_encryption_key = UserEncryptionKey.find_or_initialize_by(user_id: current_user.id)
```
[the remark this answers is not in the document]
Comment (lis2) [reply app/controllers/encrypt_controller.rb]: I think this one is fine, `find_or_create_by` is making always 2 queries, select and insert
`find_or_initialize_by` is making just select, and if doesn't exist create an instance without saving to database

```
def find_or_initialize_by(attributes, &block)
  find_by(attributes) || new(attributes, &block)
end
```

[the remark this answers is not in the document]
Comment (udan11) [reply app/controllers/encrypt_controller.rb]: Oops, you are right. :)
